### PR TITLE
Adds hinge loss function algorithm

### PR DESCRIPTION
**Pull request by PoojanSmart**
### Describe your change:

- Adding Hinge loss function which is useful loss function in Support Vector Machine (SVM) for calculating loss.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

File: machine_learning/loss_functions/hinge_loss.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import numpy as np
+
+
+def hinge_loss(y_true: np.ndarray, pred: np.ndarray) -> float:
+    """
+    Calculate the hinge loss for y_true and pred for binary classification.
+
+    Args:
+        y_true: Array of actual values (ground truth) encoded as -1 and 1.
```
Comment (tianyizheng02) [inline machine_learning/loss_functions/hinge_loss.py:24]: Please have an if-statement enforce that `y_true` must be encoded as -1 or 1 (and please raise a ValueError for invalid inputs)

File: machine_learning/loss_functions/hinge_loss.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        pred: Array of predicted values.
+
+    Returns:
+        Hinge loss
```
Comment (tianyizheng02) [inline machine_learning/loss_functions/hinge_loss.py:28]:
```suggestion
    Returns:
        The hinge loss between y_true and y_pred
```

File: machine_learning/loss_functions/hinge_loss.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+Author: Poojan Smart
+Email: smrtpoojan@gmail.com
+"""
+
+import numpy as np
+
+
+def hinge_loss(y_true: np.ndarray, pred: np.ndarray) -> float:
+    """
+    Calculate the hinge loss for y_true and pred for binary classification.
```
Comment (tianyizheng02) [inline machine_learning/loss_functions/hinge_loss.py:21]:
```suggestion
    Calculate the mean hinge loss for y_true and pred for binary classification.
```
The documentation should reflect the fact that you're averaging the hinge losses over the arrays

File: machine_learning/loss_functions/hinge_loss.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    Traceback (most recent call last):
+    ...
+    ValueError: Length of predicted and actual array must be same.
+    """
+
+    if len(y_true) != len(pred):
+        raise ValueError("Length of predicted and actual array must be same.")
+
+    intermidiate_result = 1.0 - (y_true * pred)
+    intermidiate_result[intermidiate_result < 0] = 0
```
Comment (tianyizheng02) [inline machine_learning/loss_functions/hinge_loss.py:47]:
```suggestion
    hinge_losses = np.max(0, 1 - y_true * y_pred)
```
More descriptive variable name, and you can just use `np.max`

File: machine_learning/loss_functions/hinge_loss.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    loss = np.mean(intermidiate_result)
+    return loss
```
Comment (tianyizheng02) [inline machine_learning/loss_functions/hinge_loss.py:49]:
```suggestion
    return np.mean(hinge_losses)
```

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

**Comment by cclauss** [comment]
https://github.com/TheAlgorithms/Python/pull/10637#issuecomment-1766252341 also applies here.

**Comment by PoojanSmart** [comment]
> [#10637 (comment)](https://github.com/TheAlgorithms/Python/pull/10637#issuecomment-1766252341) also applies here.

I can combine all loss function into single file, then you can review. We will do further modularization if required based on loss function category.
But before that, Is it fine if you merge this one? 
I will raise new issue and PR to compile all loss functions.

Let me know your thoughts..

File: machine_learning/loss_functions/hinge_loss.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    ...
+    ValueError: y_true can have values -1 or 1 only.
+    """
+
+    if len(y_true) != len(y_pred):
+        raise ValueError("Length of predicted and actual array must be same.")
+
+    # Raise value error when y_true (encoded labels) have any other values
+    # than -1 and 1
+    if np.array_equal(np.sort(np.unique(y_true)), np.array([-1, 1])) is False:
```
Comment (tianyizheng02) [inline machine_learning/loss_functions/hinge_loss.py:54]:
```suggestion
    if np.any((y_true != 1) & (y_true != -1)):
```
This should do the same thing (check if any entry is neither 1 or -1)

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

**Review by tianyizheng02 — APPROVED** [review]

**Pull request merged**